text/0000-force-declaration-of-ffi-functions.md

```diff
@@ -0,0 +1,96 @@
+
+# How We Teach This
+
+- The ["C FFI" chapter of the tutorial](https://tutorial.ponylang.io/c-ffi.html) should be updated to reflect this change.
+
+    In particular, the ["Calling C from Pony" section](https://tutorial.ponylang.io/c-ffi/calling-c.html) should note that a declaration is needed in order to call a FFI function, as well as introduce the `...` syntax for declaring variadic functions. Although we could expect that users that attempt to use the C FFI will be familiar with the concept of variadic functions, it wouldn't hurt to mention them, even if we don't use the "variadic" term.
+
+- All the examples that show FFI code, be it in the `examples/` folder of the compiler or on the tutorial should be changed to include mandatory declarations. These examples are located in the C FFI and [Serialisation](https://tutorial.ponylang.io/appendices/serialisation.html) chapters, and in the [examples appendix](https://tutorial.ponylang.io/appendices/examples.html).
+
+In general, the use of the C-FFI should be considered an advanced feature, as its usage can violate many of the safety properties enforced by the language. As such, not much would change for regular Pony users.
+
+# How We Test This
+
+* All currently supported platforms have common calling conventions for regular and variadic functions. The change in code generation shouldn't affect any of these platforms.
+
+* The use of the appropriate calling conventions will be handled by LLVM. We would assume that LLVM is able to do this correctly for all supported platforms.
+
+* Compiler tests can/should be written to demonstrate that programs that lack explicit FFI declarations fail to compile.
+
+* Apple ARM64 should be added as a target platform for CI when our current provider (Cirrus CI) offers instances of this type. In the meantime, the Pony team owns two machines with Apple Silicon chips that should serve as testing infrastructure.
+
+# Drawbacks
+
+* Breaks existing code that uses FFI functions.
+
+* Imposes overhead on users to declare the type of FFI functions in advance.
+
+# Alternatives
+
+* Make the compiler treat FFI functions as regular functions by default, and require explicit declaration only for variadic FFI functions.
+
+* Use an alternative syntax to let the compiler know about variadic functions at the call site, so that a declaration is not needed, such as (for example) `@printf[I32](...)(arg1, arg2)`.
+
+* Leave the compiler as it is, and do not support platforms with different calling conventions.
+
+# Unresolved questions
+
+Exact implementation details are still to be determined.
+
+If declarations are mandatory, specifying the return type at FFI call site is made redundant, and maybe should be removed from the language.
+
+Should C functions provided by the runtime (`@pony*`) be declared explicitly?
```

Comment:
For consistency, especially to avoid confusion arising from inconsistency, I would say yes.

Reply:
Agreed. Yes.

Reply:
Added to the RFC